src/main/java/org/kaazing/netx/ws/internal/WsURLConnectionImpl.java

```diff
@@ -329,39 +419,65 @@ public WsReader getReader() throws IOException {
 
     @Override
     public Writer getWriter() throws IOException {
-        if (writer == null) {
+        if (writer != null) {
+            return writer;
+        }
+
+        try {
+            stateLock.lock();
             ensureConnected();
             writer = new WsWriter(this);
+            return writer;
+        }
+        finally {
+            stateLock.unlock();
         }
-
-        return writer;
     }
 
     @Override
     public void setChallengeHandler(ChallengeHandler challengeHandler) {
+        checkConnected();
         connection.setChallengeHandler(challengeHandler);
     }
 
     @Override
     public void setConnectTimeout(int timeout) {
+        checkConnected();
         connection.setConnectTimeout(timeout);
```

Comment:
Never mind, I see it now.
Recommend changing `checkConnected()` to something like `ensureReconfigurable()` which we only support in `START` state before attempting to connect.